src/logger/src/logger.rs

```diff
@@ -1,776 +1,343 @@
-    pub static ref LOGGER: &'static Logger = {
-        set_logger(_LOGGER_INNER.deref()).expect("Failed to set logger");
-        _LOGGER_INNER.deref()
-    };
-}
-
-/// Logger representing the logging subsystem.
-// All member fields have types which are Sync, and exhibit interior mutability, so
-// we can call logging operations using a non-mut static global variable.
-pub struct Logger {
-    init: Init,
-    // Human readable logs will be outputted here.
-    log_buf: Mutex<Box<dyn Write + Send>>,
-    show_level: AtomicBool,
-    show_file_path: AtomicBool,
-    show_line_numbers: AtomicBool,
-    instance_id: RwLock<String>,
-}
-
-impl fmt::Debug for Logger {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Logger")
-            .field("init", &self.init)
-            .field("log_buf", &"?")
-            .field("show_level", &self.show_level)
-            .field("show_file_path", &self.show_file_path)
-            .field("show_line_numbers", &self.show_line_numbers)
-            .field("instance_id", &self.instance_id)
-            .finish()
-    }
-}
+/// Default instance id.
+pub const DEFAULT_INSTANCE_ID: &str = "anonymous-instance";
+/// Instance id.
+pub static INSTANCE_ID: OnceLock<String> = OnceLock::new();
+
+/// The logger.
+///
+/// Default values matching the swagger specification (`src/api_server/swagger/firecracker.yaml`).
+pub static LOGGER: Logger = Logger(Mutex::new(LoggerConfiguration {
+    target: None,
+    filter: LogFiler {
+        level: log::LevelFilter::Warn,
+    },
+    format: LogFormat {
+        show_level: false,
+        show_log_origin: false,
+    },
+}));
+
+/// Error type for [`Logger::update`].
+#[derive(Debug, thiserror::Error)]
+#[error("Failed to open target file: {0}")]
+pub struct LoggerUpdateError(pub std::io::Error);
 
 impl Logger {
-    /// Creates a new instance of the current logger.
-    fn new() -> Logger {
-        Logger {
-            init: Init::new(),
-            log_buf: Mutex::new(Box::new(sink())),
-            show_level: AtomicBool::new(true),
-            show_line_numbers: AtomicBool::new(true),
-            show_file_path: AtomicBool::new(true),
-            instance_id: RwLock::new(String::new()),
+    /// Applies the given logger configuration the logger.
+    pub fn update(&self, config: LoggerConfig) -> Result<(), LoggerUpdateError> {
+        let mut guard = self.0.lock().unwrap();
+        if let Some(level) = config.level {
+            guard.filter.level = log::LevelFilter::from(level);
         }
-    }
 
-    fn show_level(&self) -> bool {
-        self.show_level.load(Ordering::Relaxed)
-    }
-
-    fn show_file_path(&self) -> bool {
-        self.show_file_path.load(Ordering::Relaxed)
-    }
+        if let Some(log_path) = config.log_path {
+            let file = std::fs::OpenOptions::new()
+                .custom_flags(libc::O_NONBLOCK)
+                .read(true)
```

Comment:
why would we want to read from the log file?

Reply:
This is ported from the original, it also enabled `read`. I do not know why.